pkg/apiserver/apiserver.go

```diff
@@ -0,0 +1,217 @@
+	workerEventsQueueURL = regexp.MustCompile(`^/workers/(?P<id>\d+)/queue/?$`)
+	workerAllQueue       = regexp.MustCompile(`^/workers/all/queue/?$`)
+	clustersURL          = "/clusters/"
+)
+
+// New creates new HTTP API server
+func New(controller ControllerInformer, port int, logger *logrus.Logger) *Server {
+	s := &Server{
+		logger:     logger.WithField("pkg", "apiserver"),
+		controller: controller,
+	}
+	mux := http.NewServeMux()
+
+	mux.Handle("/debug/pprof/", http.HandlerFunc(pprof.Index))
+	mux.Handle("/debug/pprof/cmdline", http.HandlerFunc(pprof.Cmdline))
+	mux.Handle("/debug/pprof/profile", http.HandlerFunc(pprof.Profile))
+	mux.Handle("/debug/pprof/symbol", http.HandlerFunc(pprof.Symbol))
+	mux.Handle("/debug/pprof/trace", http.HandlerFunc(pprof.Trace))
+
+	mux.Handle("/status/", http.HandlerFunc(s.controllerStatus))
+	mux.Handle("/config/", http.HandlerFunc(s.operatorConfig))
+
+	mux.HandleFunc("/clusters/", s.clusters)
+	mux.HandleFunc("/workers/", s.workers)
+
+	s.http = http.Server{
+		Addr:        fmt.Sprintf(":%d", port),
+		Handler:     http.TimeoutHandler(mux, httpAPITimeout, ""),
+		ReadTimeout: httpReadTimeout,
+	}
+
+	return s
+}
+
+// Run starts the HTTP server
+func (s *Server) Run(stopCh <-chan struct{}, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	go func() {
+		err := s.http.ListenAndServe()
+		if err != http.ErrServerClosed {
+			s.logger.Fatalf("Could not start http server: %v", err)
+		}
+	}()
+	s.logger.Infof("Listening on %s", s.http.Addr)
+
+	<-stopCh
+
+	ctx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
+	defer cancel()
+	err := s.http.Shutdown(ctx)
+	if err == context.DeadlineExceeded {
+		s.logger.Warnf("Shutdown timeout exceeded. closing http server")
+		s.http.Close()
+	} else if err != nil {
+		s.logger.Errorf("Could not shutdown http server: %v", err)
+	}
+	s.logger.Infoln("Http server shut down")
+}
+
+func (s *Server) respond(obj interface{}, err error, w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json")
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(map[string]interface{}{"error": err.Error()})
+		return
+	}
+
+	err = json.NewEncoder(w).Encode(obj)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		s.logger.Errorf("Could not encode: %v", err)
+	}
+}
+
+func (s *Server) controllerStatus(w http.ResponseWriter, req *http.Request) {
+	s.respond(s.controller.GetStatus(), nil, w)
+}
+
+func (s *Server) operatorConfig(w http.ResponseWriter, req *http.Request) {
+	s.respond(map[string]interface{}{
+		"controller": s.controller.GetConfig(),
+		"operator":   s.controller.GetOperatorConfig(),
+	}, nil, w)
+}
+
+func (s *Server) clusters(w http.ResponseWriter, req *http.Request) {
+	var (
+		resp interface{}
+		err  error
+	)
+
+	if matches := clusterStatusURL.FindAllStringSubmatch(req.URL.Path, -1); matches != nil {
+		resp, err = s.controller.ClusterStatus(matches[0][1], matches[0][2])
```

Comment:
the numerical indexes are really hard to read/error prone. Perhaps wrap  SubexpNames in a function to extract matches by names?